File: solutions/python/markdown/1/markdown.py
import re


def parse(markdown):
    text = markdown
    for i in range(6, 0, -1):
        text = re.sub(
            r"^{} (.*?$)".format("#" * i),
            r"<h{0}>\1</h{0}>".format(i),
            text,
            flags=re.M,
        )
    text = re.sub(r"^\* (.*?$)", r"<li>\1</li>", text, flags=re.M)
    text = re.sub(r"(<li>.*</li>)", r"<ul>\1</ul>", text, flags=re.S)
    text = re.sub(r"^(?!<[hlu])(.*?$)", r"<p>\1</p>", text, flags=re.M)
    text = re.sub(r"__([^\n]+?)__", r"<strong>\1</strong>", text)
    text = re.sub(r"_([^\n]+?)_", r"<em>\1</em>", text)
    text = re.sub(r"\n", "", text)
    return text

File: solutions/go/markdown/1/markdown.go
package markdown

import (
	"fmt"
	"regexp"
	"strings"
)

// Render translates markdown to HTML
func Render(markdown string) string {
	var sb strings.Builder
	for _, line := range strings.Split(markdown, "\n") {
		sb.WriteString(toHTML(line))
	}
	return finalize(sb.String())
}

func toHTML(markdown string) string {
	for i := 1; i <= 6; i++ {
		if isHeader(markdown, i) {
			return toHeader(markdown, i)
		}
	}
	if isList(markdown) {
		return toList(markdown)
	}
	return toParagraph(markdown)
}

func isHeader(markdown string, level int) bool {
	re := regexp.MustCompile(fmt.Sprintf(`^(#){%d} `, level))
	return re.MatchString(markdown)
}

func toHeader(markdown string, level int) string {
	return fmt.Sprintf("<h%d>%s</h%d>", level, markdown[level+1:], level)
}

func isList(markdown string) bool {
	re := regexp.MustCompile(`^\* `)
	return re.MatchString(markdown)
}

func toList(markdown string) string {
	return "<ul><li>" + markdown[2:] + "</li></ul>"
}

func toParagraph(markdown string) string {
	return "<p>" + markdown + "</p>"
}

func finalize(s string) string {
	s = replaceFormatters(s, "__", "strong")
	s = replaceFormatters(s, "_", "em")
	return strings.ReplaceAll(s, "</ul><ul>", "")
}

func replaceFormatters(markdown string, mdTag string, htmlTag string) string {
	markdown = strings.Replace(markdown, mdTag, "<"+htmlTag+">", 1)
	markdown = strings.Replace(markdown, mdTag, "</"+htmlTag+">", 1)
	return markdown
}
